examples/plot_resting_correlations.py

```diff
@@ -46,14 +46,25 @@
 We want to use an appropriate color map for these data: a divergent map that
 is centered on 0, which is a meaningful transition-point as it marks the change
 from negative correlations to positive correlations.
+"""
+brain.add_data(surf_data_lh, -.7, .7, colormap="vlag", hemi='lh')
+brain.add_data(surf_data_rh, -.7, .7, colormap="vlag", hemi='rh')
 
-We'll also plot the map with some transparency so that we can see through to
-the underlying anatomy.
 """
-brain.add_data(surf_data_lh, -.7, .7, colormap="vlag", alpha=.75,
-               hemi='lh')
-brain.add_data(surf_data_rh, -.7, .7, colormap="vlag", alpha=.75,
-               hemi='rh')
+You can make the small correlations transparent by re-scaling the colormap with
+transparency and letting the rescaler know that this is a divergent color map.
+In this case the first number gives the value that should be in the center of
+the colormap, the second determines which values will be transparent and the 
+third determines the left and right edges of the new colormap.
```

Comment:
It's also common to want thresholds for divergent colormaps to effectively be:
```
[-fmax, -fmid, -fmin, 0, fmin, fmid, fmax]
```
Where `[-fmin, fmin]` is a single color (the midpoint of the colormap).

If we support this use case somehow, your use case is also automatically supported. Can you think of a good API?

One idea would be to do `center=None` default (for non-divergent) and do:

1. `add_data(...., fmin=2, fmid=4, fmax=6, center=0)` would do what I want
2. `add_data(..., fmin=0, fmid=3, fmax=6, center=0)` would do what you want

Or something equivalent and maybe nicer could be done with `divergent=False` default:

1. `add_data(..., fmin=2, fmid=4, fmax=6, divergent=True)` would do what I want
2. `add_data(..., fmin=0, fmid=3, fmax=6, divergent=True)` would do what you want

WDYT?

Reply:
I like `center=` as a way of specifying divergent colormaps.

Reply:
It would be nice if pysurfer did what seaborn does with divergent colormaps which is to clip the lookup table so that the colorbar shows the extreme values in the data while keeping the luminance midpoint in the right place.

Reply:
Also if @larsoner's original suggestion is implemented that would mean completely duplicating the `add_overlay` functionality, which is worth thinking about.

Reply:
Yes, originally I also planned to allow something like `[-fmax, -fmid, -fmin, 0, fmin, fmid, fmax]`, but settled for the easier `[c-max, c-mid, c, c+mid, c+max]`, because that's much easier to handle with a single colormap. At the latest when thinking about generating a colorbar for the version allowing for space between the positive and negative ranges I realised that this is in effect plotting two separate sets of data with two separate colormaps. I only then found add_overlay which, when called as in the example, already does what you want, just that it doesn't allow you to specify colormaps (yet?). 

As a user, I'm not actually clear about what the principle difference between add_overlay and add_data is. It appears to me that add_overlay is just a crippled version of add_data?

I also like the `center=` argument, but it makes less sense, when there is no separation between negative and positive sides, because you'd have to give both `fmin` and `center` values which should be equal. `center` could also override `fmin`, when given, which sounds fine, too.

@mwaskom I'm afraid I didn't quite understand what you mean with seaborn's handling of diverging colormaps. Intuitively, the procedure you're describing sounds like what the fmax-argument already does.

Reply:
@sbitzer described more here: https://github.com/mwaskom/seaborn/pull/1184. It's a little related to your core aim (intermediate transparency) but as I've thought about adding `center` to PySurfer to handle this aspect of diverging colormaps it might be good to do all at once.

Reply:
> As a user, I'm not actually clear about what the principle difference between add_overlay and add_data is. It appears to me that add_overlay is just a crippled version of add_data?

`add_overlay` existed first to closely reproduce tksurfer functionality. `add_data` was later added to provide a more direct mapping from an array to a heatmap overlay and gradually grew increasingly bloated with additional functionality. My preference has long been to break it up.

Reply:
`add_data` does a lot more than `add_overlay`, I also view it as having (almost) a superset of functionality.

Not sure about breaking it up, sounds like a lot of work for little gain but if someone is motivated (and can keep it working well) then it's fine by me.

I don't think we should go with just `[c-max, c-min, c, c+min, c+max]` it is less functional.

Reply:
> I don't think we should go with just [c-max, c-min, c, c+min, c+max] it is less functional.

That's not my proposal.

Reply:
That was meant as a response to @sbitzer's comment, sorry if it was ambiguous

Reply:
... but seeing as in `seaborn` the problem is a bit different than what we have here (and what I have proposed with `[-fmax, -fmid, -fmin, 0, fmin, fmid, fmax]` or a variant like `center + [...]`), can you elaborate on the proposal you'd prefer here? Would it be able to get the `[-fmax, -fmid, -fmin, 0, fmin, fmid, fmax]` functionality?

Reply:
@larsoner yes the seaborn-relevant point I'm making is just that it would be nice if diverging colormaps could satisfy both (1) there is a specified "center value" and (2) the negative and positive colorbar limits reflect the limits of the data. I understand the discussion above about whether the parameterization should be a boolean `diverging` with an implicit `center=0` or an explicit `center` parameter.

Reply:
> it would be nice if diverging colormaps could satisfy both (1) there is a specified "center value" and (2) the negative and positive colorbar limits reflect the limits of the data. 

My proposal only does (1) but not (2) IIUC. By (2) you mean supporting something like (in an extreme case) `[-2, -1.5, -1, 0, 10, 11, 12]`, right? The only way I can see to get this sort of behavior (without too much complexity) is just to allow people to specify 7 control points if they want.

Reply:
@larsoner seaborn has three parameters: `vmin`, `vmax`, and `center`. To get an opacity ramp you'd need a fourth (`vmid`).

Reply:
I'm not sure how `vmid` would work -- where would it go in terms of the control points / what would the equivalent be in the 7-point scheme?

Reply:
(My understanding of the seaborn scheme is that it would be equivalent to the seven-point scheme `[vmin, (vmin+center)/2., center, center, center, (vmax+center)/2., vmax]`)

Reply:
Here's a concrete example. You want to plot an overlay that ranges from -5 to 8. You put `add_data(..., center=0, fmin=2, fmid=4, fmax=8)`. The colorbar that you see has limits at -5 and 8. It is fully opaque in the ranges [-5, -4] and [4, 8]. It ramps down in opacity linearly between [-4, -2] and ramps up in opacity linearly between [2, 4]. It is transparent in the range [-2, 2]. (All range specs ignore details of inclusive/exclusive for now). The middle RGB value of the colormap (LUT128) corresponds to 0, the maximum RGB value of the colormap (LUT255) corresponds to 8, and an intermediate RGB value of the colormap (LUT48) corresponds to -5.

Reply:
Yeah that's fine by me -- it's equivalent to `[-8, -4, -2, 0, 2, 4, 8]` in the seven-point scheme I keep talking about -- and AFAICT already equivalent to what I proposed yesterday (right?).

However, you mentioned also wanting to have support for divergent colormaps where "(2) the negative and positive colorbar limits reflect the limits of the data.", how in this `center,fmin,fmid,fmax` 4-kwarg scheme do we get that? Maybe I don't understand what you mean by it...?

Reply:
(I assume that by (2) you mean that you want LUT0 to correspond to `vmin` and LUT255 to correspond to `vmax` in seaborn terminology, but we don't have independent/equivalents for `vmin`/`vmax` in this scheme because we force symmetry as `vmin=-fmax, vmax=fmax` effectively)

Reply:
I'll just drop this for now and code it up later if I become interested again.

Reply:
So I think the existing proposal is to use `center, fmin, fmid, fmax` as:

1. Default `center=None`, treat as sequential colormap (current behavior)
2. `center is not None` use `[center-fmax, center-fmid, center-fmin, center, fmin, fmid, fmax]` as limits (with center-fmin to fmin as LUT128, and -fmax/fmax as LUT0 and LUT255, respectively), with appropriate checks for monotonic `center <= fmin <= fmid <= fmax`

Okay with you to proceed with this @mwaskom?

Reply:
Seems fine.

Reply:
@sbitzer does this make sense to you? Do you have time to give it a shot?

Reply:
Yes, makes sense, except that I guess you mean `[center-fmax, center-fmid, center-fmin, center, center+fmin, center+fmid, center+fmax]`?

I'll see how far I get today or tomorrow and let you know of the progress.

Reply:
Yes, I think so. Thanks for taking a stab at it. A couple of minor points:

- It might be worth de-duplicating with `add_overlay` if/where possible.
- We probably want different default colormaps to be used (`cmap='auto'` by default?) depending on if `center=None` (use current default) or not (some diverging colormap). 
- It would be good if the colormap actually displays properly on the colorbar. It might, depending on how it's implemented, but we should at least check the result for e.g. `center=0, fmin=1, fmid=2, fmax=3`

@mwaskom any personal favorite diverging colormap we should use by default?

Reply:
@larsoner the `icefire` colormap I added recently is intended to be used for diverging neuroimaging visualization.

Reply:
Ok, automatically choosing a divergent colormap, when `center` is provided is a good idea. The only issue is that so far the colormap is only chosen in `add_data` while `center` would live in `scale_data_colormap`. Should I then also add the full `fmin, fmid, fmax, center` to `add_data`?

Reply:
It's okay with me, as they'll mostly get passed through (and it could save a call to `scale_data_colormap`). Would it simplify / unify `add_data` and `add_overlay` calls a bit?

Reply:
I did most of the work today, but still see a strange effect that I don't understand. I will be away the next two days and will check in my progress on Friday.